### DISCUSSION TITLE
How to use Channel.fromPath read the content from a file?

### DISCUSSION THREAD

**Discussion by DescartesM** [nextflow-io/nextflow#3963] [Q&A]
The scene is a bit special. I have a series of files in the same directory, but I don't know the number of files before and the file suffixes are not uniform; 
> for example:  a.txt, b.csv, c.yaml

My goal is to achieve a grouping effect like this
```groovy
chunks = Channel
    .fromPath('/some/path/*.txt')
    .buffer(size:2, remainder:true)
    .map { tuple(chunk_idx++, it) }

// output
[1, [a.txt, b.csv]]
[2, [c.yaml]]
```
But due to the different names and suffixes, I can not use the fromPath('/some/path/*.txt')......
So currently, I have implemented similar capabilities using shell scripts. I write these file names into a file, and use Channel.fromPath to read it.

```groovy
filename = getFiles(params.in) // the filename  content format is ['a.txt', 'b.csv', 'c.yaml']

chunks = Channel.fromPath(filename.text)
      .buffer(size:2, remainder:true)
      .map { tuple(chunk_idx++, it) }
```
However, I get the error message as following. 
> class groovyx.gpars.dataflow.expression.DataflowInvocationExpression cannot be cast to class java.nio.file.FileSystem (groovyx.gpars.dataflow.expression.DataflowInvocationExpression is in unnamed module of loader 'app'; java.nio.file.FileSystem is in module java.base of loader 'bootstrap')

I would like to know how to solve the above problem or if there are any other ways to achieve my goal, thx.

**Comment by bentsherman** [comment]
How about this:

```groovy
Channel.fromPath('/some/path/*.{txt,csv,yml}')
```

**Comment by mestia** [comment]
I encountered a similar problem where I needed to read a list of files from a file instead of using I/O-heavy filesystem globbing on a directory with numerous files. In addition, the input channel needed to be divided into pairs and sent as a list of file paths to a process.

Using splitText() alone wouldn't suffice, as it adds a newline to each element. However, the combination of splitCsv() and map{} did the trick:

```
nextflow.enable.dsl=2
filelist = file(params.filelist)
workflow {
   ch_in = Channel.fromPath(filelist).splitCsv(by:2).map{(it.flatten())}.view()
}
```
the result, in case of 3 files:
```
[testdata/file10.txt, testdata/file11.txt]
[testdata/file2.txt]
```
